Order Of Execution

DhritiShikhar opened the pull request:
fixes #508

DhritiShikhar commented on migration/sql-files/028-adds-order-to-existing-wi.sql:
```diff
@@ -0,0 +1,21 @@
+ALTER TABLE work_items ADD COLUMN position double precision;
```
order is a keyword is postgres, so i am using the word "position"

maxandersen replied:
why not something like "execution_order" so the db matches the model ?

DhritiShikhar replied:
@maxandersen how about "itemorder"?

DhritiShikhar replied:
fixes it here -> d88baa9

maxandersen commented on workitem/workitem_repository.go:
```diff
@@ -93,6 +112,110 @@ func (r *GormWorkItemRepository) Delete(ctx context.Context, ID string) error {
+			return nil, errors.NewInternalError(err.Error())
+		}
+		afterOrder, _ := strconv.ParseFloat(fmt.Sprintf("%v", afterItem.Position), 64)
+		order = afterOrder + 1000
```
why 1000 ? and what happens if this lands on something with same execution order ? 

Don't seem like we even detect that ?

DhritiShikhar replied:
@maxandersen I dont think that there could be any case where two workitems will have same order because:
[1] Order of new workitem is (highest-order-in-db + 1000) which is unique
[2] While reordering, 
Order of the reorder workitem is obtained by dividing the order of the two 'closest' workitems, so the order will always be a value between the range of those two 'closest' workitems. It will never be **equal** to the order of those two closest workitems. 

(By two 'closest' workitems, I mean the workitems below and above which the reorder workitem is going to be placed.)

maxandersen commented on workitem/workitem_repository.go:
```diff
@@ -166,10 +288,19 @@ func (r *GormWorkItemRepository) Create(ctx context.Context, typeID string, fiel
+	if err != nil {
+		return nil, errors.NewInternalError(err.Error())
+	}
+	pos = pos + 1000
```
if 1000 is so magical why not have a constant or variable for this so it is guaranteed to be consistent ?

DhritiShikhar replied:
fixes it here -> d88baa9

maxandersen commented on workitem/workitemtype.go:
```diff
@@ -43,6 +43,12 @@ const (
+
+	SystemMarkupPlainText = "PlainText"
+	SystemMarkupMarkdown  = "Markdown"
+	SystemMarkupJiraWiki  = "JiraWiki"
```
looks unrelated to the patch.

DhritiShikhar replied:
fixes it here -> d88baa9

aslakknutsen commented on migration/sql-files/028-adds-order-to-existing-wi.sql:
```diff
@@ -0,0 +1,21 @@
+ALTER TABLE work_items ADD COLUMN executionorder double precision;
```
Missing executionorder index.

DhritiShikhar replied:
fixed

aslakknutsen commented on workitem/workitem.go:
```diff
@@ -18,6 +18,8 @@ type WorkItem struct {
 	// the field values
 	Fields Fields `sql:"type:jsonb"`
+	// the position of workitem
+	Executionorder float64 `sql:"type:jsonb"`
```
Why is this type:jsonb?

kwk replied:
@aslakknutsen it now is a float64: https://github.com/almighty/almighty-core/pull/804/files#diff-bd043efe71050f0771319a56c8a73754R22

aslakknutsen commented on workitem/workitem_repository.go:
```diff
@@ -93,6 +114,114 @@ func (r *GormWorkItemRepository) Delete(ctx context.Context, ID string) error {
+	res.Fields = Fields{}
+
+	res.Executionorder = order
+	for fieldName, fieldDef := range wiType.Fields {
```
Why convert all the fields etc?  Shouldn't we just store the new executionorder?

kwk replied:
I think @DhritiShikhar no longer does so. @aslakknutsen please check the latest code if your question still holds true. @DhritiShikhar please make sure you at least answer @aslakknutsen 's question when you have changed something.

DhritiShikhar replied:
@kwk @aslakknutsen sorry. Yes i should have replied after making the changes.

aslakknutsen commented on workitem/workitem_repository.go:
```diff
@@ -166,10 +294,20 @@ func (r *GormWorkItemRepository) Create(ctx context.Context, typeID string, fiel
+		return nil, errors.NewInternalError(err.Error())
+	}
+	pos = pos + orderValue
+	fields[SystemOrder] = pos
```
SystemOrder seems to be missing a lot of the similar logic around read/write that SystemCreatedAt has.

DhritiShikhar replied:
like what?

kwk replied:
@aslakknutsen does your concern still hold true or can you please elaborate a bit?

aslakknutsen replied:
It's ok in this position, but why is SystemOrder not handled here https://github.com/DhritiShikhar/almighty-core/blob/e5b7d804073753a4737e13515df8f41493c8d783/workitem/workitem_repository.go#L333 and here https://github.com/DhritiShikhar/almighty-core/blob/e5b7d804073753a4737e13515df8f41493c8d783/workitem/workitem_repository.go#L427

Won't that just store system.order as part of the Fields jsonb when someone PATCH and not show up in WIList?

DhritiShikhar replied:
@aslakknutsen yes. I will fix this

kwk commented:
[test]

kwk commented on test/work_item_repository.go:
```diff
@@ -32,6 +32,17 @@ type WorkItemRepository struct {
+	}
+	reorderReturns struct {
+		result1 *app.WorkItem
+		result2 error
```
@DhritiShikhar does it make sense to rename `result1` to `wi` and and `result2` to `err`?

DhritiShikhar replied:
@kwk yes, The variable names should be descriptive. I fixed it in this commit -> d0dbec6

kwk commented on test/work_item_repository.go:
```diff
@@ -148,6 +159,47 @@ func (fake *WorkItemRepository) SaveReturns(result1 *app.WorkItem, result2 error
+	fake.reorderReturns = struct {
+		result1 *app.WorkItem
+		result2 error
+	}{result1, result2}
```
Why define the result struct inline and not have it as an unexported struct?

DhritiShikhar replied:
fixed

kwk commented on workitem/undoable_wi_repo.go:
```diff
@@ -65,6 +65,30 @@ func (r *UndoableWorkItemRepository) Save(ctx context.Context, wi app.WorkItem)
+
+	old := WorkItem{}
+	db := r.wrapped.db.First(&old, id)
+	if db.Error != nil {
```
Does it make sense to check for a not found error first and only then return an internal error?

```go
if db.RecordNotFound() {
	return nil, errors.NewNotFoundError("work item", id)
}
if db.Error != nil {
	return nil, errors.NewInternalError(fmt.Sprintf("could not load %s, %s", wi.ID, db.Error.Error()))
}
```

DhritiShikhar replied:
Agreed. The Internal Error should be checked before NotFounf error. Thanks for pointing it out. i fixed it here -> d0dbec6

kwk commented on workitem/workitem_repository.go:
```diff
@@ -99,6 +119,120 @@ func (r *GormWorkItemRepository) Delete(ctx context.Context, ID string) error {
+	if err != nil {
+		return nil, errors.NewBadParameterError("Type", wi.Type)
+	}
+	if beforeID != "" {
```
@DhritiShikhar Can you maybe create small functions and use them in these huge `if` blocks? The complexity of this function would dramatically decrease and maintainability increases. What do you think?

DhritiShikhar replied:
@kwk I have also added better comments and reuse code where ever possible -> d0dbec6

kwk commented on workitem/workitem_repository.go:
```diff
@@ -99,6 +119,120 @@ func (r *GormWorkItemRepository) Delete(ctx context.Context, ID string) error {
+
+	res.Executionorder = order
+
+	for fieldName, fieldDef := range wiType.Fields {
```
I wonder why you cannot just assign `wiType.Fields` to `res.Fields` and replace only the values that you need (e.g. `SystemOrder`). Then you could get rid of this for loop, no?

DhritiShikhar replied:
Yes. That's correct. Fixed it.

kwk commented on workitem_blackbox_test.go:
```diff
@@ -89,6 +91,37 @@ func TestGetWorkItemWithLegacyDescription(t *testing.T) {
+	resource.Require(t, resource.Database)
+	priv, _ := almtoken.ParsePrivateKey([]byte(almtoken.RSAPrivateKey))
+	svc := testsupport.ServiceAsUser("TestGetWorkItem-Service", almtoken.NewManagerWithPrivateKey(priv), testsupport.TestIdentity)
+	assert.NotNil(t, svc)
```
Please, make this a `require.NotNil(t, svc)` otherwise we end up having a panic when you use the `svc` variable one line below. And that shadows the actual error.

DhritiShikhar replied:
Fixed this

kwk commented on workitem_blackbox_test.go:
```diff
@@ -89,6 +91,37 @@ func TestGetWorkItemWithLegacyDescription(t *testing.T) {
+	svc := testsupport.ServiceAsUser("TestGetWorkItem-Service", almtoken.NewManagerWithPrivateKey(priv), testsupport.TestIdentity)
+	assert.NotNil(t, svc)
+	controller := NewWorkitemController(svc, gormapplication.NewGormDB(DB))
+	assert.NotNil(t, controller)
```
Same here.

kwk commented on workitem.go:
```diff
@@ -131,11 +131,11 @@ func (c *WorkitemController) Reorder(ctx *app.ReorderWorkitemContext) error {
 
 		if ctx.Payload.Position.ID == "" {
-			return jsonapi.JSONErrorResponse(ctx, errors.NewBadParameterError("missing position.id", nil))
+			return jsonapi.JSONErrorResponse(ctx, errors.NewBadParameterError("Bad Parameter position.id", nil))
```
All error messages should not be capitalized. See here: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

DhritiShikhar replied:
fixed it here -> 32a238c

kwk commented on workitem.go:
```diff
@@ -131,11 +131,11 @@ func (c *WorkitemController) Reorder(ctx *app.ReorderWorkitemContext) error {
 
 		if ctx.Payload.Position.Direction == "" {
-			return jsonapi.JSONErrorResponse(ctx, errors.NewBadParameterError("missing position.direction", nil))
+			return jsonapi.JSONErrorResponse(ctx, errors.NewBadParameterError("Bad parameter position.direction", nil))
```
Same here: https://github.com/golang/go/wiki/CodeReviewComments#error-strings

kwk commented on design/media_types.go:
```diff
@@ -62,9 +62,10 @@ var meta = a.Type("workItemListResponseMeta", func() {
 	a.Description("Position represents the ID of the workitem above which the to-be-reordered workitem(s) should be placed")
-	a.Attribute("above", d.String, "ID of the workitem above which the to-be-reordered workitem(s) should be placed")
+	a.Attribute("id", d.String, "ID of the workitem above which the to-be-reordered workitem(s) should be placed")
+	a.Attribute("direction", d.String, "position of workitem")
```
I think the description is a bit off, don't you think? How can a `direction` be a `position` at the same time?

kwk replied:
If this really is a direction parameter, it might make sense to define it like this:

```go
a.Attribute("direction", d.String, "...", func() {
    a.Enum("above", "below")
})
```

Alternatively if the string MUST NOT be empty, you can define it like so:

```go
a.Attribute("direction", d.String, "...", func() {
    a.MinLength(1)
})
```

Then you don't need to check for empty strings in the controller. The same is true (at least for `a.MinLength(1)`) for the `id` attribute.

DhritiShikhar replied:
fixed it here -> 32a238c

kwk commented on workitem.go:
```diff
@@ -126,10 +126,18 @@ func (c *WorkitemController) Update(ctx *app.UpdateWorkitemContext) error {
 			return jsonapi.JSONErrorResponse(ctx, errors.NewBadParameterError("missing payload element in request", nil))
 		}
 
+		if ctx.Payload.Position.ID == "" {
```
Please see my comment above about the `a.Enum("above", "below")` or `a.MinLength(1)`. By applying any of the solutions you can get rid of this check.

kwk commented on workitem.go:
```diff
@@ -126,10 +126,18 @@ func (c *WorkitemController) Update(ctx *app.UpdateWorkitemContext) error {
+			return jsonapi.JSONErrorResponse(ctx, errors.NewBadParameterError("missing position.id", nil))
+		}
+
+		if ctx.Payload.Position.Direction == "" {
```
Please see my comment above about the `a.Enum("above", "below")` or `a.MinLength(1)`. By applying any of the solutions you can get rid of this check.

kwk commented on workitem/workitem_repository.go:
```diff
@@ -119,27 +121,77 @@ func (r *GormWorkItemRepository) Delete(ctx context.Context, ID string) error {
+//      FindSecondItem returns the value above which reorder item has to be placed
+func (r *GormWorkItemRepository) FindSecondItem(order *float64, secondItem string, WorkItemId string) (*uint64, *float64, error) {
+	Item := WorkItem{}
+	if secondItem == below {
```
I'd make this a `switch` with a `default` case.

DhritiShikhar replied:
@kwk what would the `default` case have? Its either below or above

kwk replied:
Then you must ensure that `secondItem` can only contain "below" or "above". If you do that, you don't need a default case.

DhritiShikhar replied:
makes sense. I fixed this here -> 5362fc0

kwk commented on workitem/workitem_repository.go:
```diff
@@ -119,27 +121,77 @@ func (r *GormWorkItemRepository) Delete(ctx context.Context, ID string) error {
+// If position.direction == "below", then
+//	FindItemItem returns the value below which reorder item has to be placed
+//      FindSecondItem returns the value above which reorder item has to be placed
+func (r *GormWorkItemRepository) FindSecondItem(order *float64, secondItem string, WorkItemId string) (*uint64, *float64, error) {
```
Why not use the `app.WorkItemReorderPosition` for `secondItem` and `WorkItemID`? Btw. check `make check-go-format` for hints on variable names. The output is really handy.

DhritiShikhar replied:
@kwk that's because I just need the order of secondItem. Also, I can reuse the tx errors; otherwise I will have to write them twice in Reorder() function

kwk commented on workitem/workitem_repository.go:
```diff
@@ -119,27 +121,77 @@ func (r *GormWorkItemRepository) Delete(ctx context.Context, ID string) error {
+	} else if secondItem == above {
+		// Finds the item above which reorder item has to be placed
+		tx := r.db.Where("executionorder > ?", order).Order("executionorder", true).Last(&Item)
+		if tx.RecordNotFound() {
```
what about other errors?

DhritiShikhar replied:
Yes. tx.Error should be handled.

kwk commented on workitem/workitem_repository.go:
```diff
@@ -119,27 +121,77 @@ func (r *GormWorkItemRepository) Delete(ctx context.Context, ID string) error {
+	if secondItem == below {
+		// Finds the item below which reorder item has to be placed
+		tx := r.db.Where("executionorder < ?", order).Order("executionorder desc", true).Last(&Item)
+		if tx.RecordNotFound() {
```
what about other errors?

DhritiShikhar replied:
Yes. tx.Error should be handled.

DhritiShikhar commented:
@aslakknutsen As per our discussion, I fixed all these cases:
[1] Handle the error when same request is sent twice. -> This is done by checking the ID of workitem
[2] If the order of workitem is not changes, then "save" function should not be executed.
[3] Add another parameter -> payload.position.direction to specify the reorder workitem should be placed "above" or "below" payload.position.id. This is done to handle cases such as pagination

kwk commented on workitem/workitem_repository.go:
```diff
@@ -119,27 +121,77 @@ func (r *GormWorkItemRepository) Delete(ctx context.Context, ID string) error {
+	if tx.Error != nil {
+		return nil, errors.NewInternalError(err.Error())
+	}
+	if tx.RecordNotFound() {
```
this must be placed before line 167

DhritiShikhar replied:
yes. Fixing it.

kwk commented on workitem/workitem_repository.go:
```diff
@@ -119,27 +121,77 @@ func (r *GormWorkItemRepository) Delete(ctx context.Context, ID string) error {
 	if tx.Error != nil {
 		return nil, errors.NewInternalError(err.Error())
 	}
+	if tx.RecordNotFound() {
```
this must be checked first

DhritiShikhar commented:
Sample request payload:

curl -i -X PATCH localhost:8080/api/workitems/reorder -H "Authorization: Bearer ..." -d '{"data": [{"type": "workitems", "attributes":{"version":0}, "id": "5"}], "position": {"id":"2", "direction":"above"}}'

// In case of pagination where only 10 workitems are loaded in the first page, direction: "below" would help
curl -i -X PATCH localhost:8080/api/workitems/reorder -H "Authorization: Bearer ..." -d '{"data": [{"type": "workitems", "attributes":{"version":0}, "id": "5"}], "position": {"id":"10", "direction":"below"}}'

DhritiShikhar commented:
@aslakknutsen fixed the 3 things as per our discussion:
[1] Handle the case when same request is sent twice. -> Compare the ids
[2] Donot proceed the transaction if the workitem when order is not changed
[3] Add another parameter in payload payload.Position.Direction which can hold value "above" or "below" to specify where the reorder workitem is going to be placed.

DhritiShikhar commented:
Sample Request payload:

curl -i -X PATCH localhost:8080/api/workitems/reorder -H "Authorization: Bearer ..." -d '{"data": [{"type": "workitems", "attributes":{"version":0}, "id": "5"}], "position": {"id":"2", "direction":"above"}}'


// handle pagination
curl -i -X PATCH localhost:8080/api/workitems/reorder -H "Authorization: Bearer ..." -d '{"data": [{"type": "workitems", "attributes":{"version":0}, "id": "3"}], "position": {"id":"10", "direction":"below"}}'

DhritiShikhar commented:
Video of reordering workitems -> https://www.youtube.com/watch?v=pPpQ_TET7Ps

kwk commented on workitem/workitem_repository.go:
```diff
@@ -230,15 +250,17 @@ func (r *GormWorkItemRepository) Reorder(ctx context.Context, position *app.Work
 			// Item is placed at last position
 			order = *belowItemOrder + float64(orderValue)
-		} else if *aboveItemId == res.ID {
+		} else if *aboveItemId == string(res.ID) {
```
I think this can never be `true` at least for how you intended it to work. Check this example:

```go
package main

import (
	"fmt"
)

func main() {
	zero := "0"
	t := string(0)
	fmt.Printf("zero = '%+v'\nt = '%+v'\n", zero, t)
        // Outputs:
        // zero = '0'
        // t = ''
}
```

You should use [`FormatUint`](https://golang.org/pkg/strconv/#FormatUint) instead to convert an unsigned integer into a string.

DhritiShikhar commented:
@kwk As per our discussion, I fixed following:
* In design, added enum for position.Direction to hold values above and below only
* Uses FormatUInt() instead of uint64() in Reorder() function in workitem_repository.go
* Removes redundant code in FindSecondItem() function
* Adds proper comment in Reorder test cases
* Checks for order of returned workitems greater than or less than in Reorder test case
* Modifies CalculateOrder() to be one liner

kwk commented on controller/workitem_blackbox_test.go:
```diff
@@ -130,18 +130,32 @@ func TestReorderWorkItem(t *testing.T) {
 	assert.Equal(t, *result3.Data.ID, *reordered1.Data[1].ID)
 
+	// Return error if order of reordered workitem is greater than the workitem position.id
+	if (reordered1.Data[0].Attributes["order"].(float64) > result1.Data.Attributes["order"].(float64)) || reordered1.Data[1].Attributes["order"].(float64) > result1.Data.Attributes["order"].(float64) {
```
I'd make two asserts out of this one line:

```go
assert.True(t, reordered1.Data[0].Attributes["order"].(float64) <= result1.Data.Attributes["order"].(float64))
assert.True(t, reordered1.Data[1].Attributes["order"].(float64) <= result1.Data.Attributes["order"].(float64))
```

DhritiShikhar replied:
I don't think i should use "<=" here. Because order should never be same for two workitems. Just "<"

kwk commented on controller/workitem_blackbox_test.go:
```diff
@@ -130,18 +130,32 @@ func TestReorderWorkItem(t *testing.T) {
-	_, reordered1 := test.ReorderWorkitemOK(t, nil, nil, controller, &payload2)
-	require.Len(t, reordered1.Data, 2)
+
+	_, reordered1 := test.ReorderWorkitemOK(t, nil, nil, controller, &payload2) // Returns the workitems which have reordered
```
I just noticed that you're only testing the "positive" case in which everything is supposed to be all good (`ReorderWorkitemOK`). But according to the reorder action definition there are probably also negative tests possible which you can put into a separate test function.

kwk commented on design/media_types.go:
```diff
@@ -63,7 +63,9 @@ var meta = a.Type("workItemListResponseMeta", func() {
 	a.Attribute("id", d.String, "ID of the workitem above which the to-be-reordered workitem(s) should be placed")
-	a.Attribute("direction", d.String, "position of workitem")
+	a.Attribute("direction", d.String, func() {
+		a.Enum("above", "below")
```
This is exactly what I meant. Well done!

kwk commented on workitem/workitem_repository.go:
```diff
@@ -136,40 +135,29 @@ func (r *GormWorkItemRepository) CalculateOrder(above, below *float64) float64 {
+	} else {
+		ItemId := strconv.FormatUint(Item.ID, 10)
+		return &ItemId, &Item.Executionorder, nil
+	}
```
There's no return after this last `else` block so you can just omit the `else` all together and just return without further nesting.

DhritiShikhar replied:
fixed

kwk commented on controller/workitem_blackbox_test.go:
```diff
@@ -101,6 +108,129 @@ func TestGetWorkItemWithLegacyDescription(t *testing.T) {
+	assert.Equal(t, *result3.Data.ID, *reordered1.Data[1].ID)
+
+	// Return error if order of reordered workitem is greater than the workitem position.id
+	if (reordered1.Data[0].Attributes["order"].(float64) > result1.Data.Attributes["order"].(float64)) || reordered1.Data[1].Attributes["order"].(float64) > result1.Data.Attributes["order"].(float64) {
```
What about splitting this into different asserts for easier readability? See https://github.com/almighty/almighty-core/pull/804#discussion_r103488430

DhritiShikhar replied:
oops. I missed that comment. I will fix it

kwk commented on controller/workitem_blackbox_test.go:
```diff
@@ -101,6 +108,129 @@ func TestGetWorkItemWithLegacyDescription(t *testing.T) {
+	assert.Equal(t, *result4.Data.ID, *reordered2.Data[0].ID)
+
+	// Return error if order of reordered workitem is less than the workitem position.id
+	if reordered2.Data[0].Attributes["order"].(float64) < result4.Data.Attributes["order"].(float64) {
```
What about splitting this into different asserts for easier readability? See https://github.com/almighty/almighty-core/pull/804#discussion_r103488430

kwk commented on controller/workitem_blackbox_test.go:
```diff
@@ -101,6 +108,129 @@ func TestGetWorkItemWithLegacyDescription(t *testing.T) {
 }
 
+// TestSuccessReorder is positive test which tests successful reorder by providing valid input
+func TestSuccessReorder(t *testing.T) {
```
Please stick to the convention of building test function names: Test + What + How (+ Expected).  In other words you should have have named this `TestReorderWorkitemOK`.

DhritiShikhar replied:
fixed here -> dab6ead

kwk commented on controller/workitem_blackbox_test.go:
```diff
@@ -101,6 +108,129 @@ func TestGetWorkItemWithLegacyDescription(t *testing.T) {
+}
+
+// TestFailReorderMissingReorderItem is negative test which tests unsuccessful reorder by providing invalid input
+func TestFailReorderMissingReorderItem(t *testing.T) {
```
Please rename test to `TestReorderWorkitemBadRequest`

DhritiShikhar replied:
fixed here -> dab6ead

kwk commented on controller/workitem_blackbox_test.go:
```diff
@@ -101,6 +108,129 @@ func TestGetWorkItemWithLegacyDescription(t *testing.T) {
+}
+
+// TestFailReorderMissingPosition is negative test which tests unsuccessful reorder by providing invalid input
+func TestFailReorderWorkItemNotFound(t *testing.T) {
```
Please rename to `TestReorderWorkitemNotFound`

DhritiShikhar replied:
fixed here -> dab6ead

kwk commented on controller/workitem.go:
```diff
@@ -130,13 +130,6 @@ func (c *WorkitemController) Reorder(ctx *app.ReorderWorkitemContext) error {
 			return jsonapi.JSONErrorResponse(ctx, errors.NewBadParameterError("missing payload element in request", nil))
 		}
 
-		if ctx.Payload.Position.ID == "" {
```
It's good that you have omitted those tests! :+1:

DhritiShikhar commented:
@kwk thank you for being patient with me and your time and efforts in reviewing this PR.

kwk commented on workitem/workitem_repository.go:
```diff
@@ -98,6 +120,159 @@ func (r *GormWorkItemRepository) Delete(ctx context.Context, ID string) error {
+	default:
+		return &wi, nil
+	}
+	fmt.Println(order)
```
Remove this line please

DhritiShikhar replied:
fixed -> 802b584

aslakknutsen commented on workitem/workitem.go:
```diff
@@ -18,6 +18,8 @@ type WorkItem struct {
 	// the field values
 	Fields Fields `sql:"type:jsonb"`
+	// the position of workitem
+	Executionorder float64
```
ExecutionOrder ?

DhritiShikhar replied:
@aslakkuntsen why not Executionorder? I didnt keep it "order" because in postgres "order" is a keyword. and i am trying to avoid that word.

DhritiShikhar replied:
@aslakknutsen @kwk why not Executionorder? I didnt keep it "order" because in postgres "order" is a keyword. and i am trying to avoid that word.

kwk replied:
@DhritiShikhar I think what @aslakknutsen meant is that you should use an uppercase `O`, right? So `Executionorder` becomes `ExecutionOrder`.

DhritiShikhar replied:
@kwk @aslakknutsen oh ok. I fixed it here -> 2326c52

aslakknutsen commented on workitem/workitemtype.go:
```diff
@@ -43,6 +43,8 @@ const (
 	SystemStateResolved   = "resolved"
 	SystemStateClosed     = "closed"
+
+	SystemOrder = "order"
```
Move this up where the other Fields are defined.

DhritiShikhar replied:
Fixed it here -> 8c1e086

aslakknutsen commented:
There should be a link on the WorkItem List API 'reorder': 'url' that the UI can react to to indicate if the reorder feature is supported for this list or not.

kwk commented:
> There should be a link on the WorkItem List API 'reorder': 'url' that the UI can react to to indicate if the reorder feature is supported for this list or not.

We can add this in another PR I'd say.

DhritiShikhar commented:
There should be a link on the WorkItem List API 'reorder': 'url' that the UI can react to to indicate if the reorder feature is supported for this list or not.

DhritiShikhar commented:
@aslakknutsen @kwk created a issue for that -> https://github.com/almighty/almighty-core/issues/910

aslakknutsen commented on workitem/workitemtype.go:
```diff
@@ -26,6 +26,7 @@ const (
 	SystemCreatedAt           = "system.created_at"
 	SystemIteration           = "system.iteration"
 	SystemArea                = "system.area"
+	SystemOrder               = "order"
```
Why is this not system.order?

kwk replied:
I think this is for no particular reason. @DhritiShikhar please, rename as @aslakknutsen suggested.

DhritiShikhar replied:
@aslakknutsen fixed it here -> 5966205

aslakknutsen replied:
@DhritiShikhar No you didn't.

DhritiShikhar replied:
I made a mistake. I will fix it.

aslakknutsen commented on controller/workitem_blackbox_test.go:
```diff
@@ -101,6 +108,125 @@ func TestGetWorkItemWithLegacyDescription(t *testing.T) {
+	payload1.Data.Attributes[workitem.SystemTitle] = "Reorder Test WI"
+	payload1.Data.Attributes[workitem.SystemState] = workitem.SystemStateClosed
+	_, result1 := test.CreateWorkitemCreated(t, svc.Context, svc, controller, &payload1)
+	defer test.DeleteWorkitemOK(t, nil, nil, controller, *result1.Data.ID)
```
Any reason you're not using the DeleteEntities thing here?

kwk replied:
@aslakknutsen I think it is because the test in not part of a suite but  sure. @DhritiShikhar you could have something like this

```go
db := gormapplication.NewGormDB(DB)
defer cleaner.DeleteCreatedEntities(db)
```

DhritiShikhar replied:
@aslakknutsen @kwk yes i will fix this

DhritiShikhar replied:
@aslakknutsen @kwk fixed it here -> 700819f

aslakknutsen commented on controller/workitem_blackbox_test.go:
```diff
@@ -101,6 +108,125 @@ func TestGetWorkItemWithLegacyDescription(t *testing.T) {
+	_, result4 := test.CreateWorkitemCreated(t, svc.Context, svc, controller, &payload1)
+	defer test.DeleteWorkitemOK(t, nil, nil, controller, *result4.Data.ID)
+	_, result5 := test.CreateWorkitemCreated(t, svc.Context, svc, controller, &payload1)
+	defer test.DeleteWorkitemOK(t, nil, nil, controller, *result5.Data.ID)
```
DeleteEntities?

DhritiShikhar replied:
@aslakknutsen fixed it here -> 700819f

aslakknutsen commented on controller/workitem_blackbox_test.go:
```diff
@@ -92,6 +98,7 @@ func TestGetWorkItemWithLegacyDescription(t *testing.T) {
 
 	_, updated := test.UpdateWorkitemOK(t, nil, nil, controller, *wi.Data.ID, &payload2)
 	assert.NotNil(t, updated.Data.Attributes[workitem.SystemCreatedAt])
+	assert.NotNil(t, updated.Data.Attributes["order"])
```
workitem.SystemOrder ?

kwk replied:
@DhritiShikhar use `assert.NotNil(t, updated.Data.Attributes[workitem.SystemOrder])` here instead as @aslakknutsen suggested.

DhritiShikhar replied:
@kwk fixed it here -> 5966205

aslakknutsen commented on workitem/workitem_repository.go:
```diff
@@ -17,10 +17,15 @@ import (
 type WorkItemRepository interface {
 	Load(ctx context.Context, ID string) (*app.WorkItem, error)
 	Save(ctx context.Context, wi app.WorkItem) (*app.WorkItem, error)
+	Reorder(ctx context.Context, position *app.WorkItemReorderPosition, wi app.WorkItem) (*app.WorkItem, error)
```
Why pass the rest API position object in here?

DhritiShikhar replied:
@aslakknutsen To calculate the reorder item's position (https://github.com/almighty/almighty-core/pull/804/files#diff-55f2da878ed2fcb42e5373010f02ff92R210)

I am passing rest API workitem object. So, why not rest API position object?

aslakknutsen replied:
Because you're passing REST objects down into the DB layer. You are now tightly coupling the layout of the REST API with the domain/db layer. We had the same with app.WorkItem vs app.WorkItem2. And we have the same with WorkItemLinks etc etc. It creates some couplings that cause some very odd behavior. Like we can't store a WorkItemLink without having a HTTP Request Context.

The Reorder function could easily be:

    Reorder(ctx contextContext, direction Direction, target uuid.UUID, wi app.WorkItem)

DhritiShikhar replied:
@aslakknutsen You are right. I will make the required changes.`

aslakknutsen commented on controller/workitem_blackbox_test.go:
```diff
@@ -41,6 +41,11 @@ import (
 	"github.com/stretchr/testify/suite"
```
Missing test cases for 'move to top' and 'move to bottom'?

DhritiShikhar replied:
@aslakknutsen For 'move to top' and 'move to bottom', there wont be any change in the request payload. Its just that it would execute a different if control statement -> if Item.ID == 0. The tests that I have written move an Item to the top of list and bottom of the list. So, I have handled this case as well in the tests and I don't think there is a need to write tests explicitly for that.

aslakknutsen replied:
It's not about whether or not it executes the same code. A test should capture the expected behavior in different scenarios. 

Maybe I missed it, but I couldn't see a Test that shows that ID = "0" is handled in any special way at all. 

And why is it ID = "0" which sounds like a fairly magic string as oppose to ID = nil?

DhritiShikhar replied:
_And why is it ID = "0" which sounds like a fairly magic string as oppose to ID = nil?_

The ID = "0" and not "nil" because if transaction fails because of Internal Error then tx.Error returns Item.ID == "nil". I need to differentiate between the case where transaction fails and the case where record is not found.

aslakknutsen replied:
You know that based on err.RecordNotFound?

DhritiShikhar replied:
@aslakknutsen I also had doubts about the same. My concern was if I get tx.NotFoundError, It is an ERROR, and it should not be ignored and proceeded.

```
tx = r.db.Where("execution_order > ?", order).Order("execution_order", true).Last(&Item)

If tx.RecordNotFound() {
                  return Item.ID, nil, tx.Error
}
```

@aslakknutsen @kwk does this look like a proper way to handle the case?

DhritiShikhar replied:
> A test should capture the expected behavior in different scenarios.


@aslakknutsen I will seperate the test cases and handle one scenario per test case. There are following scenarios:
[1] Reorder item between two items:
            [a] using above
            [b] using below
[2] Reorder item to top
[3] Reorder item to bottom
[4[ Reorder multiple workitems

DhritiShikhar replied:
@kwk is it okay if I name the tests TestReorderAbove(), TestReorderBelow(), TestReorderTop(), TestReorderBottom(), TestReorderMultipleWorkItems()?

Because now I am using test.ReorderWorkitemOK(..) in many test functions. So, I cannot name all my functions TestReorderWorkItemOK().

kwk replied:
@DhritiShikhar just name them `test.ReorderWorkitemAboveOK`, `test.ReorderWorkitemTopOK`, and so forth.

DhritiShikhar replied:
@kwk done

aslakknutsen replied:
> [2] If Item is present AND if I get tx.Error. Then it also returns a tx.RecordNotFound()

Why does it return RecordNotFound on 'other errors'?

DhritiShikhar replied:
@aslakknutsen I was wrong. I have modified my code ->

https://github.com/almighty/almighty-core/pull/804/files#diff-55f2da878ed2fcb42e5373010f02ff92R157

Does this make sense?

If not, please tell me how i can fix it.

kwk commented on workitem/workitem_repository.go:
```diff
@@ -229,16 +230,17 @@ func (r *GormWorkItemRepository) Reorder(ctx context.Context, position *app.Work
-			return nil, errors.NewNotFoundError("work item", position.ID)
+			return nil, errors.NewNotFoundError("work item", targetID)
 		}
+		fmt.Println(">>>>>>>>>>>>>>>>>>>>>>")
```
Please remove this line!

kwk commented on workitem/workitem_repository.go:
```diff
@@ -195,7 +196,7 @@ func (r *GormWorkItemRepository) Reorder(ctx context.Context, position *app.Work
 		return nil, errors.NewNotFoundError("work item", wi.ID)
 	}
-	if tx.Error != nil {
+	if err := tx.Error; err != nil {
```
This is not really needed for just one line but it sure works.

aslakknutsen commented on workitem/workitem_repository.go:
```diff
@@ -17,10 +17,15 @@ import (
 type WorkItemRepository interface {
 	Load(ctx context.Context, ID string) (*app.WorkItem, error)
 	Save(ctx context.Context, wi app.WorkItem, modifierID uuid.UUID) (*app.WorkItem, error)
+	Reorder(ctx context.Context, direction string, targetID string, wi app.WorkItem, modifierID uuid.UUID) (*app.WorkItem, error)
```
Could you make 'direction' a Type. That way the API is clear that it needs .g. a DirectionType which has 2 possible constants above or below, not just an arbitrary string.

kwk replied:
@aslakknutsen Would have been nice to have this for the work item ID as well ;) . That is also just a arbitrary string that needs to be convertible into a number. But that's a different story ;)

DhritiShikhar replied:
will fix this

aslakknutsen replied:
@kwk That should be moved to UUID :)

DhritiShikhar replied:
@aslakknutsen Isn't a.Enum("above", "below") enough? -> https://github.com/almighty/almighty-core/pull/804/files#diff-7d67dc444801d745cdc248547dcce7c7R70

DhritiShikhar replied:
@aslakknutsen should it be something like this? ->

```
// Direction represents the direction of the workitem according to which the to-be-reordered workitem(s) should be placed
var Direction = a.Type("workItemReorderDirection", func() {
	a.Description("Direction represents the direction of the workitem according to which the to-be-reordered workitem(s) should be placed")
	a.Attribute("above", d.String, "Above should be used to place the reorder workitem(s) above workitem with id equal to position.id. ", func() {
		a.Enum("above")
	})
	a.Attribute("below", d.String, "Below should be used to place the reorder workitem(s) below workitem with id equal to position.id.", func() {
		a.Enum("below")
	})
})
```

kwk replied:
> @aslakknutsen Isn't a.Enum("above", "below") enough?

The enum is fine for the outward facing API side but internally, I think @aslakknutsen wants it to be a type like this:

```go
type DirectionType string
const (
  Above DirectionType = "above"
  Below DirectionType = "below"
)
```

Then you can use the `direction DirectionType` instead of the `direction string`.

@aslakknutsen correct?

DhritiShikhar replied:
@kwk @aslakknutsen fixed this

aslakknutsen commented on workitem/workitemtype.go:
```diff
@@ -27,7 +27,9 @@ const (
 	SystemArea                = "system.area"
-	SystemCodebase            = "system.codebase"
+	SystemOrder               = "system.order"
+
```
Remove empty line sp codebase aligns with the other fields.

DhritiShikhar replied:
i will fix this

aslakknutsen commented on controller/workitem_blackbox_test.go:
```diff
@@ -148,6 +154,188 @@ func (s *WorkItemSuite) TestCreateWI() {
+
+// TestReorderTop is positive test which tests successful reorder by providing valid input
+// This case reorders one workitem -> result2 and places it to the top of the list
+func (s *WorkItemSuite) TestReorderWorkitemTopOK() {
```
It's a bit unclear to me why/how this test tests the Move To Top behavior?

DhritiShikhar replied:
@aslakknutsen There are two workitems in list-> result1 and result2. It moves result2 above result1.

aslakknutsen replied:
Yes, that's the problem. It moved Result2 below Result1 specifically. How is that different from what TestReorderWorkitemAboveOK|TestReorderWorkitemBelowOK does?

The purpose of this test must be to express the feature of 'move to top'. The top of the list. Not just above something specific.

I think an ok test case here would be:

insert A, B, C
move C to Top  (_not_ before A)
check C before A

DhritiShikhar replied:
Right now, the request payload looks like this:

Assume you want to move workitem 5 to top i.e. above workitem 1, then:

curl -i -X PATCH localhost:8080/api/workitems/reorder -H "Authorization: Bearer " -d '{"data": [{"type": "workitems", "attributes":{"version":0}, "id": "5"}], "position": {"id":"1", "direction":"above"}}'

When you said -> 

> move C to Top (not before A)

Should the request payload also change to depict that an item is moved to top and not above some item (in this case 1)?

kwk replied:
@aslakknutsen @DhritiShikhar do we all agree that

  "move to top"

is a synonym for

  "move above work item that is **currently** the top most in the list"

The only problem I see is the time at which the top most work item is determined. If this isn't done in one transaction there potentially may be problems, right? For example, it shouldn't be determined on the client but on the core side.

My suggestion would be to develop something like an `api/workitems/reorder/totop` and `api/workitems/reorder/tobottom` endpoint and only pass in the work item that you want to move. Then figure out the top/bottom-most work item in the repository instead of providing it together with the request.

@aslakknutsen would that help with your concerns?

@DhritiShikhar said that she plans to work the night (her time) in order to catch up with you @aslakknutsen.

DhritiShikhar replied:
@aslakknutsen @kwk i think i might have to write a function-> LoadFirstItem from DB in workitem/workitem_repository.go. Something similar to this -> https://github.com/almighty/almighty-core/pull/804/files#diff-55f2da878ed2fcb42e5373010f02ff92R99

DhritiShikhar replied:
@kwk How do we decide that we need a new API endpoint? I mean, why create a new api endpoint for this? Why not handle it in the same API endpoint workitems/reorder?

aslakknutsen replied:
@DhritiShikhar We spoke about this a long time ago. The UI won't know what is 'the first' work item when it's on page 4, so the API needs a way to move X to the 'top of the list'. And same with Bottom of the list. 

You support this behavior currently(i think), mind you a bit strange, by 'if belowItemId == 0" (aboveItemId). But the test doesn't reflect this.

aslakknutsen replied:
@DhritiShikhar I think I miss read some code here, thinking the if belowItemId was handling the 'move to top of list' part, but that's only handling the case where there is no wi to move this between. Which leaves the feature of 'move this to top/bottom of list' missing as far as I can tell?

DhritiShikhar replied:
Yes you are correct. But doesn't it mean the same thing? Move to top means that there is no second item. FindSecondItem() would return belowItemID = 0 @aslakknutsen

DhritiShikhar replied:
@aslakknutsen as per discussion, I have to support Move to Top and Move To Bottom which I thought was supported already but it was not a perfect implementation. To fix this, I would do the following:
[1] Modify request payload to clearly mention move to top or bottom of the list. 
[2] Write function in workitem_repository which would give the workitem with highest order (for move to top) and workitem with lowest order (for move to bottom). 
[3] Make changes in Reorder(...) function in workitem_repository accordingly. 
[4] Fix test cases according to that. 

And as per discussion, no new API endpoint is needed.

DhritiShikhar replied:
@aslakknutsen thanks for pointing this out. I realised what a loop hole it was while fixing it.

aslakknutsen commented on controller/workitem_blackbox_test.go:
```diff
@@ -148,6 +154,188 @@ func (s *WorkItemSuite) TestCreateWI() {
+
+// TestReorderBottom is positive test which tests successful reorder by providing valid input
+// This case reorders one workitem -> result1 and places it to the bottom of the list
+func (s *WorkItemSuite) TestReorderWorkitemBottomOK() {
```
It's a bit unclear to me why/how this test tests the Move To Bottom behavior?

DhritiShikhar replied:
There are two workitems in list-> result1 and result2. It moves result1 below result2

kwk commented on workitem/workitem_repository.go:
```diff
@@ -18,14 +18,19 @@ import (
-const above = "above"
-const below = "below"
+
+type DirectionType string
```
Why re-defined it here again?

aslakknutsen replied:
I assume you mean why re-define it in the blackbox_test. This is actually the correct location :)

kwk replied:
@aslakknutsen yes, sorry. I meant: Why redefined this in the blackbox test?

kwk commented on workitem/workitem_repository.go:
```diff
@@ -114,6 +141,160 @@ func (r *GormWorkItemRepository) Delete(ctx context.Context, workitemID string,
+		if *aboveItemId == "0" {
+			// Item is placed at last position
+			order = *belowItemOrder + float64(orderValue)
+		} else if *aboveItemId == string(res.ID) {
```
Is this string conversion still in your PR?

DhritiShikhar replied:
i will fix this

kwk commented on workitem/workitem_repository.go:
```diff
@@ -95,6 +97,43 @@ func (r *GormWorkItemRepository) Load(ctx context.Context, ID string) (*app.Work
+// returns NotFoundError, ConversionError or InternalError
+func (r *GormWorkItemRepository) LoadTopWorkitem(ctx context.Context) (*app.WorkItem, error) {
+	res := WorkItem{}
+	r.db = r.db.Debug()
```
Please just use r.db and skip the Debug.

DhritiShikhar replied:
I forgot to remove this line.

kwk replied:
@DhritiShikhar if you've fixed it. Please put the commit here

DhritiShikhar replied:
@kwk f7f1483

kwk commented on workitem/workitem_repository.go:
```diff
@@ -95,6 +97,43 @@ func (r *GormWorkItemRepository) Load(ctx context.Context, ID string) (*app.Work
+func (r *GormWorkItemRepository) LoadTopWorkitem(ctx context.Context) (*app.WorkItem, error) {
+	res := WorkItem{}
+	r.db = r.db.Debug()
+	tx := r.db.Order("execution_order desc").Last(&res)
```
Ordering and then taking the last item is a waste of time and could take a long time. Since we don't need the ordering anyway, we can instead just use the `max` function instead:

```sql
SELECT *
FROM work_items
WHERE
  execution_order = (SELECT max(execution_order) FROM work_items);
```

DhritiShikhar replied:
@kwk i also had the same idea initially but the reason i didn't do so was gorm doesnot support aggregate functions.

kwk replied:
@DhritiShikhar what about this?

```go
db := r.db.Model(WorkItem{})
query := fmt.Sprintf("execution_order = (SELECT max(execution_order) FROM %[1]s)",
	workitem.WorkItem{}.TableName(),
)
res = WorkItem{}
db = db.Where(query).First(&res)
```

See [here](https://github.com/almighty/almighty-core/blob/7255e80325f5ca9f9a241564bf003a4d2e492a40/workitem/link/type_repository.go#L266) for an example of more complicated queries.

DhritiShikhar replied:
@kwk okay. I will fix this

kwk replied:
@DhritiShikhar if you've fixed it. Please put the commit here

DhritiShikhar replied:
@kwk d9c97ca

kwk commented on workitem/workitem_repository.go:
```diff
@@ -95,6 +97,43 @@ func (r *GormWorkItemRepository) Load(ctx context.Context, ID string) (*app.Work
+	r.db = r.db.Debug()
+	tx := r.db.Order("execution_order desc").Last(&res)
+	if tx.RecordNotFound() {
+		return nil, nil
```
Here you return a `nil` error but in `LoadBottomWorkitem` you return the `tx.Error`. I'd say you should return a `errors.NewNotFoundError` in both cases, no?

DhritiShikhar replied:
`errors.NewNotFoundError(..) ` requires `ID` of workitem as well. In this case, there is no specific workitem ID that I am passing. So, is `return nil, tx.Error` good enough?

kwk replied:
@DhritiShikhar It's definitively better than returning no error.

DhritiShikhar replied:
@kwk i will fix this

kwk replied:
@DhritiShikhar if you've fixed it. Please put the commit here

DhritiShikhar replied:
@kwk d9c97ca

kwk replied:
Now you do`db = db.Where(query).First(&res)` but don't check for `db.Error` any longer. That wasn't my intention. Is that on purpose?

kwk commented on workitem/workitem_repository.go:
```diff
@@ -95,6 +97,43 @@ func (r *GormWorkItemRepository) Load(ctx context.Context, ID string) (*app.Work
+// returns NotFoundError, ConversionError or InternalError
+func (r *GormWorkItemRepository) LoadBottomWorkitem(ctx context.Context) (*app.WorkItem, error) {
+	res := WorkItem{}
+	tx := r.db.Order("execution_order").Last(&res)
```
Same applies here: Don't sort, use `max`.

kwk replied:
Btw. you do `tx := r.db.Order("execution_order desc").Last(&res)` also in `LoadTopWorkitem`. Shouldn't one find the minimum and one find the maximum execution order value instead of both finding the maximum execution order?

kwk replied:
In fact it looks like the two functions `LoadTopWorkitem` and `LoadBottomWorkitem` are almost identical.

DhritiShikhar replied:
@kwk for topmost workitem -> db.Order("execution_order **desc**").Last(&res) 
For bottom most workitem -> db.Order("execution_order").Last(&res)

kwk replied:
Ah, good point. But the ordering isn't necessary here like I've mentioned before. Use `Max` instead.

aslakknutsen replied:
@kwk max/min can operate purely on the index(the index is stored ordered so it either reads forward or backward), but the same is true for the current order usage. This only becomes a problem when you do order with where clauses to e.g. limit execution order within a space.

Another 'optimization' is; Why are you reading the whole WorkItem, are you not just interested in the execution order?

@DhritiShikhar Could you create two issues for this?

DhritiShikhar replied:
@aslakknutsen 

[1] I have created a issue for Limiting Order Of Execution within a space -> https://github.com/almighty/almighty-core/issues/972

[2] Regarding second optimization, the reason that I am reading the whole workitem and not just the execution_order is because I am interested in the workitem ID also -> to handle the case where the request is made more than once. Do you think a better way would be return ID as well as execution_order?

kwk replied:
@DhritiShikhar if you've fixed it. Please put the commit here

kwk commented on workitem/workitem_repository.go:
```diff
@@ -248,24 +287,49 @@ func (r *GormWorkItemRepository) Reorder(ctx context.Context, direction Directio
+		// if direction == "top", place the reorder item at the topmost position. Now, the reorder item has the highest order in the whole list
+		res, err := r.LoadTopWorkitem(ctx)
+		if err != nil {
+			return nil, errs.WithStack(err)
```
Please use `Wrapf` instead.

DhritiShikhar replied:
fixed this

kwk replied:
@DhritiShikhar if you've fixed it. Please put the commit here

DhritiShikhar replied:
@kwk f7f1483

kwk commented on workitem/workitem_repository.go:
```diff
@@ -248,24 +287,49 @@ func (r *GormWorkItemRepository) Reorder(ctx context.Context, direction Directio
+		// if direction == "bottom", place the reorder item at the bottom most position. Now, the reorder item has the lowest order in the whole list
+		res, err := r.LoadBottomWorkitem(ctx)
+		if err != nil {
+			return nil, errs.WithStack(err)
```
Please use `Wrapf` instead.

DhritiShikhar replied:
fixed this

kwk replied:
@DhritiShikhar if you've fixed it. Please put the commit here

DhritiShikhar replied:
@kwk f7f1483

DhritiShikhar commented on controller/workitem.go:
```diff
@@ -130,6 +130,45 @@ func (c *WorkitemController) Update(ctx *app.UpdateWorkitemContext) error {
+func (c *WorkitemController) Reorder(ctx *app.ReorderWorkitemContext) error {
+	currentUserIdentityID, err := login.ContextIdentity(ctx)
+	if err != nil {
+		jerrors, _ := jsonapi.ErrorToJSONAPIErrors(goa.ErrUnauthorized(err.Error()))
```
I will fix this.

kwk replied:
@DhritiShikhar if you've fixed it. Please put the commit here.

DhritiShikhar replied:
@kwk f7f1483

aslakknutsen commented on design/media_types.go:
```diff
@@ -68,6 +68,7 @@ var position = a.Type("workItemReorderPosition", func() {
 	})
 	a.Attribute("direction", d.String, "Direction of the place of the reorder workitem. Above should be used to place the reorder workitem(s) above workitem with id equal to position.id. Below should be used to place the reorder workitem(s) below workitem with id equal to position.id", func() {
 		a.Enum("above", "below", "top", "bottom")
+		a.Description("Above places the reorder item(s) above the reference workitem. Below places the reorder item(s) below the reference workitem. Top places the reorder workitem(s) at the Topmost position of the list. Bottom places the reorder item(s) at the bottom of the list.")
```
I meant describe Top/Bottom in the Description that was part of the Attribute already.

DhritiShikhar replied:
ok. I will fix this

DhritiShikhar commented:
[test]

DhritiShikhar commented:
[test]

aslakknutsen commented on workitem/workitem_repository.go:
```diff
@@ -116,6 +181,185 @@ func (r *GormWorkItemRepository) Delete(ctx context.Context, workitemID string,
+		if err != nil {
+			return nil, errors.NewNotFoundError("work item", *targetID)
+		}
+		if *belowItemId == "0" {
```
There is no test case that test the two special cases where you want to move something 'before' the first, or 'after' the last wi? And you can't determine the first being the wi with id 0

DhritiShikhar replied:
@aslakknutsen didn't we agree that moving something 'above' first workitem is just not the best implementation? We should remove this if condition and check for nil instead, what do you think?

aslakknutsen commented on workitem/workitem_repository.go:
```diff
@@ -116,6 +181,185 @@ func (r *GormWorkItemRepository) Delete(ctx context.Context, workitemID string,
+// If direction == "below", then
+//	FindFirstItem returns the value below which reorder item has to be placed
+//      FindSecondItem returns the value above which reorder item has to be placed
+func (r *GormWorkItemRepository) FindSecondItem(order *float64, secondItemDirection DirectionType, WorkItemId string) (*string, *float64, error) {
```
Argument WorkItemId is never used

DhritiShikhar replied:
I have fixed this.

DhritiShikhar commented:
Reorder above:

`curl -i -X PATCH localhost:8080/api/workitems/reorder -H "Authorization: Bearer ..." -d '{"data": [{"type": "workitems", "attributes":{"version":0}, "id": "3"}], "position": {"id":"7", "direction":"above"}}'`

Reorder below:

`curl -i -X PATCH localhost:8080/api/workitems/reorder -H "Authorization: Bearer ..." -d '{"data": [{"type": "workitems", "attributes":{"version":0}, "id": "1"}], "position": {"id":"8", "direction":"below"}}'`

Reorder to bottom of the list:

`curl -i -X PATCH localhost:8080/api/workitems/reorder -H "Authorization: Bearer ..." -d '{"data": [{"type": "workitems", "attributes":{"version":0}, "id": "2"}], "position": {"direction":"bottom"}}'`

Reorder to top of the list:

`curl -i -X PATCH localhost:8080/api/workitems/reorder -H "Authorization: Bearer ..." -d '{"data": [{"type": "workitems", "attributes":{"version":0}, "id": "9"}], "position": {"direction":"top"}}'`

DhritiShikhar commented:
[test]

DhritiShikhar commented:
[tes]

DhritiShikhar commented:
[test]

DhritiShikhar commented:
@aslakknutsen True. I think we need that for loop back and it should be like this:

```
for fieldName, fieldDef := range wiType.Fields {
		if fieldName == SystemCreatedAt || fieldName == SystemUpdatedAt || fieldName == SystemOrder {
			continue
		}
		fieldValue := wi.Fields[fieldName]
		var err error
		res.Fields[fieldName], err = fieldDef.ConvertToModel(fieldName, fieldValue)
		if err != nil {
			return nil, errors.NewBadParameterError(fieldName, fieldValue)
		}
	}

```

DhritiShikhar commented:
[test]

aslakknutsen commented on workitem/workitem_repository.go:
```diff
@@ -538,7 +538,7 @@ func (r *GormWorkItemRepository) listItemsFromDB(ctx context.Context, criteria c
 		db = db.Limit(*limit)
 	}
-	db = db.Select("count(*) over () as cnt2 , *")
+	db = db.Select("count(*) over () as cnt2 , *").Order("execution_order")
```
@DhritiShikhar So what did you check that made you conclude ASC is the correct order?

DhritiShikhar replied:
@aslakknutsen i checked this through Vikram's board view implementation.

aslakknutsen replied:
@DhritiShikhar That's not what I mean. 

In a list of 100 work items, where they have execution_order from 1 to 100. And I'm getting a List of these work items with a page size of 20. 

What is the execution_order values on page 1 ?

DhritiShikhar replied:
100, 99, 98, 97, and so on till.. 91

aslakknutsen replied:
Right. And based on the ordering you are doing in the SQL above, ascending, the result from GET /api/workitems is:

```
postgres=# SELECT count(*) over () as cnt2 , fields->>'system.title', execution_order FROM "work_items"  WHERE "work_items".deleted_at IS NULL AND (('true')) ORDER BY execution_order LIMIT 22;                  
 cnt2 | ?column? | execution_order 
------+----------+-----------------
    3 | Test 1   |            1000
    3 | Test 2   |            2000
    3 | Test 3   |            3000
(3 rows)
```

To get the correct ordering you need to do: 

```go
Order("execution_order desc")
```

Agree?

DhritiShikhar replied:
@aslakknutsen yes i agree. Fixed this.

DhritiShikhar commented:
As per @aslakknutsen 's suggestion, the naming convention could be better. Cases for "DirectionAbove" and "DirectionBelow" needs to be reversed.

To make this change, I need to fix the test cases too.

kwk commented:
@aslakknutsen does this commit really fix your concerns about the conversion problem? a00f045

kwk commented:
To summarize what @DhritiShikhar, @michaelkleinhenz, @sbose78, @baijum, and I agreed on and found out in our call:

* The core returns the work item list to the UI ordered by `execution_order` in descending order and the UI renders it as is.
* The natural order when doing a `SELECT` in postgres is ascending order. 

To compensate for this difference in perspectives between UI and core, we decided that we have to have some translation going on between those components.

The most easiest and maintainable fix is to translate on core side:

* above -> below
* below -> above
* top -> bottom
* bottom -> top

@DhritiShikhar did this by changing the values of the `DirectionAbove`, `DirectionBelow`, `DirectionTop`, `DirectionBottom` parameters because then the tests will still operate. The fix is isolated and will retrieve a comment to track the issue of finding a more permanent solution for fixing this.

DhritiShikhar commented:
[test]

baijum commented:
[test]

aslakknutsen commented on design/media_types.go:
```diff
@@ -30,6 +30,7 @@ var workItem = a.MediaType("application/vnd.workitem+json", func() {
 	a.Description("A work item hold field values according to a given field type")
 	a.Attribute("id", d.String, "unique id per installation")
 	a.Attribute("version", d.Integer, "Version for optimistic concurrency control")
+	a.Attribute("ExecutionOrder", d.Number, "Order of the workitem in the list")
```
?!? What are you doing?

app.WorkItem.Fields[system.Order]
workitem.WorkItem.ExecutionOrder

The problem was, on workitem.WorkItemRepository.Create() you managed to store app.WorkItem.Fields[system.Order] as part of workitem.WorkItem.Fields[system.Order]. It should not be stored in workitem.WorkItem.Fields[system.Order] because it's already stored in workitem.WorkItem.ExecutionOrder, which you did manage to remove from workitem.WorkItem.Fields[system.Order] when you did a workitem.WorkItemRepository.Save()

aslakknutsen replied:
How does the current behavior work like 'CreatedAt' and 'UpdatedAt' as we spoke about?

DhritiShikhar replied:
@aslakknutsen yes, i realise that its a mistake. I am fixing it in my next commit. And as we discussed, execution_order should be placed like version. In response, it should be within fields, but in database it should be stored seperately.

aslakknutsen replied:
Keep in mind there are 3 objects in play here

app.WorkItem2.Data.Attributes[system.Order]
app.WorkItem.Fields[system.Order]
workitem.WorkItem.ExecutionOrder

DhritiShikhar replied:
okay.

DhritiShikhar replied:
@aslakknutsen -> https://github.com/DhritiShikhar/almighty-core/pull/1/files

DhritiShikhar commented:
@aslakknutsen @kwk please review -> https://github.com/DhritiShikhar/almighty-core/pull/1/files

DhritiShikhar commented:
@kwk i merged the pr against my own branch as per @aslakknutsen 's suggestion

kwk commented on workitem/workitem_repository.go:
```diff
@@ -193,12 +192,12 @@ func (r *GormWorkItemRepository) FindSecondItem(order *float64, secondItemDirect
-
 	case DirectionAbove:
 		// Finds the item above which reorder item has to be placed
+		tx = r.db.Where("execution_order < ?", order).Order("execution_order desc", true).Last(&Item)
```
Please improve this query to not sort all WIs with an execution order smaller than `?`. Something like this should do it:

```sql
SELECT *
FROM work_items
WHERE id = (SELECT max(id) FROM work_items WHERE execution_order < ?))
```

But make sure you use `WorkItem{}.TableName()` instead of just `work_items`.

DhritiShikhar replied:
@kwk i had tried doing that before as you had suggested. 

So, this is the query:

When direction is "**below**" in request payload
 _curl -i -X PATCH localhost:8080/api/workitems/reorder -H "Authorization: Bearer ..." -d '{"data": [{"type": "workitems", "attributes":{"version":0}, "id": "5"}], "position": {"id":"2", "direction":"below"}}'_->

then this SQL query to find second item:
**tx = r.db.Where(fmt.Sprintf("id = (SELECT max(execution_order) FROM %s WHERE (execution_order < ?))", WorkItem{}.TableName()), order).First(&Item)**

Similarly, When direction is **above**:
**tx = r.db.Where(fmt.Sprintf("id = (SELECT min(execution_order) FROM %s WHERE (execution_order > ?))", WorkItem{}.TableName()), order).First(&Item)**

When i run the same query in psql, it runs perfectly. But when i try running it through the application, it sometimes returns "0" even when it is not an edge case. M stuck here and trying to find out why this is happening.

kwk replied:
@DhritiShikhar as discussed on BJNs: I made a mistake (`max(id)`) that you fixed (`max(execution_order)`) but then you compared it against an `id` field which is wrong.

DhritiShikhar replied:
@kwk thank you for helping me out

kwk commented on workitem/workitem_repository.go:
```diff
@@ -193,12 +192,12 @@ func (r *GormWorkItemRepository) FindSecondItem(order *float64, secondItemDirect
+
+	case DirectionBelow:
+		// Finds the item below which reorder item has to be placed
 		tx = r.db.Where("execution_order > ?", order).Order("execution_order", true).Last(&Item)
```
Same thing here (see https://github.com/almighty/almighty-core/pull/804/commits/740b0049093ed6aabb40581508772320b7144da0#r106606847)

michaelkleinhenz commented:
@aslakknutsen the UI has some open issues on the integration, but none of them is breaking, so thie PR can be merged. We are looking at the UI things now. Merging this would actually help there, in fact.

aslakknutsen commented:
@michaelkleinhenz Sure, assuming this does what it should. So far it hasn't.

aslakknutsen commented:
@DhritiShikhar 

What's the intended behavior:
3, 2, 1  move 2 below 1

1 having execution_order 0

What I see in the database is:
after the move both 1 and 2 has order 0

DhritiShikhar commented:
@aslakknutsen how did 1 get the execution_order 0?

DhritiShikhar commented:
@aslakknutsen I could not re-create the issue.. If i send -> move id 2 "below" id 1 .. it gives me order "500" which is correct. I don't know how did you get execution_order 0 for id 2.. Can you please help recreate the issue?
